### PR TITLE
Add a spec for "Snippets"

### PR DESCRIPTION
**Pull request by zadjii-msft**
This specs out a lot of plans for snippets. We've already got these in the sxnui as "tasks", but we can do so very much more. 

This spec is a few years old now, but it's time for it to get promoted out of my draft branch. 

References: 
* #1595
* #7039
* #3121
* #10436
* #12927
* #12857
* #5790
* #15845

### CODE REVIEW

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+
+If the `save` subcommand is ran without any other subcommands, the Terminal will
+imply the `-w 0` arguments, to attempt to send this action to the current
+Terminal window. (unless of course, `-w` was manually provided on the
+commandline). When run with other subcommands, then the action will just be ran
+in the same window as all the other subcommands.
+
+> [!NOTE] In other team discussions, we've considered initially merging this
+> subcommand as `x-save`, where `x-` implies "experimental". We may want to use
+> that to merge [#16513] while we wait for this spec to be approved.
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:316]: Personally, I'd prefer we just call it `save` from the start or we wait for the spec to merge. We'd probably have to keep support for `x-save` forever for legacy purposes and that just sounds icky, but that's just my 2 cents.

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+
+### UI/UX Design
+
+For the most part, we'll be using the [Suggestions UI] to display tasks to the
+user. This is a text cursor-relative UI surface that can quickly display actions
+to the user, in the context of what they're working on.
+
+The following are some examples from VsCode, Warp. These are meant to be
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:324]:
```suggestion
The following are some examples from VsCode and Warp. These are meant to be
```

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+illustrative of what these menus already look like in the wild:
+
+![](img/vscode-tasks-000.gif)
+
+![](img/warp-workflows-000.gif)
+
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:330]:
```suggestion
![VS Code demo of tasks](img/vscode-tasks-000.gif)

![Warp demo of workflows](img/warp-workflows-000.gif)
```

\#accessibility

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+> [!NOTE] In other team discussions, we've considered initially merging this
+> subcommand as `x-save`, where `x-` implies "experimental". We may want to use
+> that to merge [#16513] while we wait for this spec to be approved.
+
+### UI/UX Design
+
+For the most part, we'll be using the [Suggestions UI] to display tasks to the
+user. This is a text cursor-relative UI surface that can quickly display actions
+to the user, in the context of what they're working on.
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:322]: I'm down to make this a part of the Suggestions UI, but I still have longstanding concerns about running out of space (after all, a use case is that this'll be used to store long/complex commands).

Can we add a parameter to have this experience through the command palette?

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+illustrative of what these menus already look like in the wild:
+
+![](img/vscode-tasks-000.gif)
+
+![](img/warp-workflows-000.gif)
+
+A prototype of saving a command directly to the user's settings, then invoking
+it via the suggestions UI
+
+![](img/save-command.gif)
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:334]:
```suggestion
![use `wtd -w 0 save foo bar --baz` to save the command, then invoke the saved command via the suggestions UI](img/save-command.gif)
```

This one's really just me feeling weird about images not having alt text

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+
+A prototype of reading tasks from the CWD
+
+![](img/tasks-from-cwd.gif)
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:338]:
```suggestion
![navigate into a folder with a .wt.json file, view that it's populated, then invoke a "clean & build solution" task via suggestions UI](img/tasks-from-cwd.gif)
```

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+
+
+## Tenets
+
+<table>
+
+<tr><td><strong>Compatibility</strong></td><td>
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:345]: `wt save` also helps alleviate some of this annoyance with JSON. And if we can figure out a story with how this fits into the settings UI, that'd be nice too.

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 8]
```diff
@@ -0,0 +1,526 @@
...
+
+</td></tr>
+
+</table>
+
+### Other potential issues
+
+Something like `wt save ping 8.8.8.8 > foo.txt` isn't going to work the way
+users want. The shell is gonna get the first crack at parsing that commandline,
+and is going to try and redirect the _output of `wt`_ to `foo.txt`.
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:401]: Hmm... Are there any other potential commandline args/subcommands we'd be interested for snippets? Here's a few ideas to spitball:

### Idea 1
`wt save --output <save output to this file>`
`wt save --snippet <command to save>`

`--output` defaults to nothing
`--snippet` is optional

### Idea 2
`wt save output <path>`
`wt save snippet <command>`

This makes them mutually exclusive (essentially, the subcommand becomes `save-output` and `save-snippet`).

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 9]
```diff
@@ -0,0 +1,526 @@
...
+
+If we've got a `.wt.json` in a given directory, should we be dynamically
+adding/removing other settings too? Profiles?
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:404]: Running down the list of things in the settings model:
- Profile: this could be cool. I'm thinking a simple Razzle profile would be nice.
- actions: done
- color schemes: eh, not really useful imo

I think profiles is worth considering, but as a future consideration. We'd have to think about discoverability here.

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 10]
```diff
@@ -0,0 +1,526 @@
...
+
+## Implementation Plan
+
+### 🐣 Crawl
+* [ ] The Command Palette and Suggestions UI need to be able to display both the
+  command name and a tooltip for the comment
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:410]: For accessibility, remember to set the help text for the comment. Tooltips aren't really accessible since they may disappear.

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 11]
```diff
@@ -0,0 +1,526 @@
...
+  - This will need to be reconciled with [#7039], which tracks displaying
+    non-localized names in the command palette
+  - It won't be a TeachingTip, since those are an unmitigated disaster. But we
+    can just fake it with another text box.
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:414]: Aww man. It looked great with the teaching tip! What happened? We can't do something like "press F2 to open help text"?

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 12]
```diff
@@ -0,0 +1,526 @@
...
+  * `--settings`: Manually save to the settings file?
+  * `--profile`: save to this profile???? Not sure if this is actually possible.
+    Maybe with the `WT_SESSION_ID` env var to figure out which profile is in use
+    for the pane with that ID
+* Longer workflows might be better exposed as notebooks. We've already got a
+  mind to support [markdown in a notebook-like
+  experience](https://github.com/microsoft/terminal/issues/16495) in the
+  Terminal. For longer scripts that may need rich markup between commands, that
+  will likely be a better UX.
+* For what it is worth, [Warp] uses .yaml files for their "workflows". As an
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:469]: Two slightly different thoughts on importing the Warp YAML:
1. `wt import` is tracked [here](https://github.com/microsoft/terminal/issues/10083). Yeah, it's a different request, but figured it'd be nice to be aware of this.
2. We've discussed [importing a color scheme via the settings UI](https://github.com/microsoft/terminal/issues/8639) in the past. Your idea here is basically the same thing as that, but for snippets. Presumably, snippets will have their own page in the settings UI? Do they need one.

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 13]
```diff
@@ -0,0 +1,526 @@
...
+  [`clone_all_repos_in_org.yaml`](https://github.com/warpdotdev/workflows/blob/main/specs/git/clone_all_repos_in_org.yaml).
+  We may want to straight up just seamlessly support that syntax as well.
+  Converting them to WT-compatible json is fairly trivial [[1](#footnote-1)].
+  Furthermore, the commands are all licensed under Apache 2.0, which means they
+  can be easily consumed by other OSS projects and shared with other developers.
+  This leads us to the next future consideration:
+* Discoverability will be important. Perhaps the actions page could have a
+  toggle to immediately filter to "snippets"? Which then also displays some text
+  like "Tip: save snippets directly from the commandline with
+  `wt save <the commandline>`".
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:480]: Will Snippets be in the Settings UI? To what degree? I feel like that's missing from this spec.

You mention it a little bit here. I'd like it if we could commit to some of it being in the Settings UI in the walk/run phase. I'm concerned about discoverability and having it in the SUI would help, at least a little bit.

**Review by carlos-zamora — CHANGES_REQUESTED** [review]

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 14]
```diff
@@ -0,0 +1,526 @@
...
+This was already added in [#16185]. These will allow third-party developers to
+create apps which add additional snippets to the Terminal. These will require
+app developers to add `id`s to each action they add in this way. Users can then
+bind that action `id` to a keybinding, if they so choose.
+
+Case in point:
+https://github.com/abduvik/just-enough-series/tree/master/courses/docker+docker-compose.
+Something like that should be able to be easily added directly to the Terminal.
+
+### Snippets pane
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:214]: oh derp I didn't add a gif here _or_ in #17330

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 15]
```diff
@@ -0,0 +1,526 @@
...
+Users should be able to save commands as snippets directly from the commandline.
+Consider: you've just run the command that worked the way you need it to. You
+shouldn't have to open the settings to then separately copy-paste the command in
+to save it. It should be as easy as <kbd>Up</kbd>, <kbd>Home</kbd>, `wt save `,
+<kbd>Enter</kbd>.
+
+The exact syntax as follows:
+
+This will be powered by a `saveSnippet` action behind the scenes. After running
+this command, a toast will be presented to the user to indicate success/failure.
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:292]: oh yea, I didn't call that out since all the `wt` subcommands that take params are _all_ basically just an action with args behind the scenes

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 16]
```diff
@@ -0,0 +1,526 @@
...
+  command. If omitted, then the parameter will be left blank, and the command
+  will use the auto-generated "Send input:..." name in menus.
+* `--description,-d`: The description to optionally assign to the command.
+* `commandline`: The commandline to save as the `input` of the `sendInput` action.
+
+If the `save` subcommand is ran without any other subcommands, the Terminal will
+imply the `-w 0` arguments, to attempt to send this action to the current
+Terminal window. (unless of course, `-w` was manually provided on the
+commandline). When run with other subcommands, then the action will just be ran
+in the same window as all the other subcommands.
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:312]: That's mostly just to avoid us popping a _new_ terminal window, to do something when the Terminal is already running. Otherwise, by default, we'll do a `new-tab` first (creating a new window)

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 1 above)
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:316]: > We'd probably have to keep support for x-save forever

ah, the whole point of `x-` is "this is experimental and will _definitely_ go away in a future build"

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 11 above)
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:414]: Oh I faked it with basically a hand-rolled teaching tip, and that _isn't_ an unmitigated disaster. I'll dig up the gif

Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:414]: https://github.com/microsoft/terminal/pull/14939#issuecomment-1948254094

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:214]: snippets-pane

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 1 above)
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:316]: Also, I've got a mind to merge #16513 while this spec is in review.

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 4 above)
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:322]: 1. For things that are in the settings file directly (and fragments), those will already be in the command palette
2. I'm not super opposed to the idea, but maybe more as a future consideration. 
3. I'm thinking the Snippets pane will be the easiest for much, much longer commands (especially with the text previews there)
4. I also do have a mind to bring back the "ghost text" previews (to show the text that'll be typed)

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 9 above)
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:404]: oh yeah this is something that's definitely top-of-mind for me right now: https://github.com/microsoft/devhome/issues/3005

Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:404]: I'm specifically calling that thought out as a future consideration, since that's too much of a shower thought for now.

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 13 above)
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:480]: I mean, they're just actions fundamentally. They're in the actions editor just the same as any other action. When I talked with @PankajBhojwani about this, we agreed that a dedicated Snippets page is basically just a _second_ actions page.  Maybe we could add a button to filter that list specifically to snippets... but there's a lot of work to be done on that page in general. Didn't feel like it was necessary to call out.

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 12 above)
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:469]: (i've linked this to the other two threads)

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 11 above)
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:414]: Added a link to that PR

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 17]
```diff
@@ -0,0 +1,526 @@
...
+use as well. To support this, the Terminal will automatically look for a
+`.wt.json` file in any directories that are parents of the CWD of the shell, and
+load actions from that file as if it were a fragment extension as well. That
+will start with the `startingDirectory` for any new panes created. If the user
+has shell integration configured to tell the Terminal about the CWD, then we'll
+refresh that list as the user changes directories.
+
+* In `Terminal.Settings.Model`, we will store a cached map of path->actions.
+  * that if multiple panes are all in the same CWD, they don't need to
+    individually re-read the file from disk and regenerate that part of the map.
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:251]: In my current implementation, I don't hot-reload that file. I'll warm-reload it (if you update the settings, we'll flush the cached actions and re-load them). 

* From purely a theoretical standpoint, I don't think we should try to read the file from disk literally every time the user asks for suggestions. On paper that sounds bad
* From just local testing, reading the file from disk every time was totally fine. 
* We could maybe optimize this a little by also caching the last `LastWriteTime` of the file, and only loading it if the timestamp is different. 🤷

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 18]
```diff
@@ -0,0 +1,526 @@
...
+        ]
+    },
+    "name": "Upload to project board",
+    "description": "Sync all our issues and bugs that have been triaged and are actually on the backlog to the big-ol project",
+},
+```
+
+This is slightly more maintainable. But now, a user could set
+`"waitForSuccess": true`, and if any part of the script fails, then the rest of
+it won't be sent to the shell.
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:201]: I'm considering punting this whole parameter to Future Considerations. It's already dependent on Shell Integration being enabled, and that makes the whole section hard to parse. You've got to wrap basically everything that mentions it with "Assuming the user also has shell integration enabled, then {{this parameter does something}}".

I called it `waitForSuccess`, because it would literally wait for the FTCS_D to indicate the command ended (and the exit code), before sending more text. `stopIfFailed` reads a little different. That wording seems to imply that it would only stop sending commands if it saw an error. (but in the default case, it's not going to be seeing anything, because the average user doesn't have SI enabled).

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 8 above)
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:401]: I did drop notes in Future Considerations for this:

> * We may want to add additional params to the `save` subcommand in the future,
>   to configure where the snippet is saved:
>   * `--local`: Save to the `.wt.json` in the CWD, if there is one (or create
>     one)
>   * `--parent`: Save to the `.wt.json` in the first ancestor of the CWD, if
>     there is one. Otherwise create one here.
>   * `--settings`: Manually save to the settings file?
>   * `--profile`: save to this profile???? Not sure if this is actually possible.
>     Maybe with the `WT_SESSION_ID` env var to figure out which profile is in use
>     for the pane with that ID
>     * This would probably require p

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 13 above)
Comment (PankajBhojwani) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:520]: Agreed, I think we can come back to SUI discoverability once the actions page has had its overhaul! The idea of a dedicated snippets page feels messy to me since those send input actions would also be in the actions page

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 19]
```diff
@@ -0,0 +1,588 @@
...
+
+The exact syntax as follows:
+
+This will be powered by a `saveSnippet` action behind the scenes. After running
+this command, a toast will be presented to the user to indicate success/failure.
+The schema for this action is as follows:
+
+```json
+"SaveSnippetAction": {
+  "description": "Arguments corresponding to a saveSnippet Action",
```
Comment (DHowett) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:296]: I don't think this needs to be parseable from JSON. It's an implementation detail.

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 20]
```diff
@@ -0,0 +1,588 @@
...
+
+> [INFO!]
+>
+> TODO!: Let's make sure to discuss the filename. It doesn't need to be
+> `.wt.json`. That seemed to match things like `.clang-format`, `.vsconfig`,
+> etc, but then also included the extension. However, node projects just use
+> `package.json` without the leading `.` for storing per-project commands.
+> Perhaps `.wt.json` is the worst of both? The best of both? `.wtconfig`?
+
+Users may also want to leave snippets in the root of their repo, for others to
```
Comment (DHowett) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:240]: Review discussion about format changes

Comment (DHowett) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:240]: Does this require TermControl to have knowledge of TerminalSettingsModel, or jsoncpp? How does it generate (internally) SendInput actions? Is this a layering violation?

**Review by DHowett — CHANGES_REQUESTED** [review]

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 20 above)
Comment (DHowett) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:240]: "No"

**Comment by zadjii-msft** [comment]
Notes from spec review:

* [ ] We should have a toast to trust things in the CWD, ala vscode trusting workspaces (store in state)
* [ ] Remove the whole section on wt.json in the command palette. f it. it can just be in the sxnui
* [ ] Specific discussion on syntax for this file: JSON? fragment JSON? some other subset?
* [ ] Do it like the snippets json below[1]
* [ ] "The exact syntax as follows" is on wrong line
* [ ] Just don't parse the `saveSnippet` action, cause what's even the point
* [ ] clarify and elaborate on the "children supersede parents"
* [ ] future consideration: snippets pane can prompt to add new snippet
* [ ] leave `wt save` experimental for a while, probably until there's a dedicated save dialog. Commandline string escaping and quote parsing is neigh impossible
* [ ] at least toast if we fail to load `.wt.json` for sxnui, snippets pane can also just display the error
* [ ] v2: can we just use `tasks.json` syntax from vscode? 
  * left as future consideration, cause those are FAR more powerful than what we're going to do now
* [ ] do we need to do due diligence that we're reading arbitrary files off the file system?
* [ ] Unanimous consent vote: `.wt.json` it is. 

[1]: the json: 
```json
  {
    "$version": "1.0.0",
    "snippets":
    [
        {
            "input": "bx\r" 
            "name": "Build project",
            "description": "Build the project in the CWD"
        },
  ```

**Comment by zadjii-msft** [comment]
[b999c51](https://github.com/microsoft/terminal/pull/17329/commits/b999c51e1b4801253704a0d9e5ce02f84c7f34ae) has all the feedback from the review yesterday

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 21]
```diff
@@ -255,26 +287,30 @@ refresh that list as the user changes directories.
...
     invoked.
-* The Command Palette is trickier, since it binds directly to the action map.
-  We'd need to be able to freely modify that map at runtime, which might be
-  prohibitively annoying. <!-- * If we want these actions to show up in the
-  Command Palette, we'll need to:
-  * We'll stash these actions in the action map as they're loaded. (in
-    `Terminal.Settings.Model`)
-  * We'll need to be able to dynamically remove them at runtime from the map (in
-    `Terminal.Settings.Model`) -->
+* These snippets will not be included in the Command Palette.
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:296]: So, I know we reached this conclusion in the review yesterday. Can we include why here and if that's a permanent solution or just one that's deprioritized?

For reference, I _do_ want them to be in the command palette (or at least have a mode in the command palette dedicated to snippets). But I still don't fully understand why we're entirely dropping it. So trying to be respectful and open-minded here.

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 22]
```diff
@@ -490,14 +564,31 @@ various tools. Stored publicly on a GitHub repo (a la the winget-pkgs repo).
...
-vein as `local` is.)
+Snippets. That could easily be done as another suggestion source.
+
+#### Profiles in `.wt.json`
+
+If we've got a `.wt.json` in a given directory, should we be dynamically
+adding/removing other settings too? Wouldn't profiles also make sense? Take for
+example, the Terminal repo. We've got a PowerShell build environment and a CMD
+one. What if we could drop two profiles in the `.wt.json` file, with the
+`commandline`'s set up to call those scripts as needed?
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:575]: IIRC there's also been discussion before about snippets that only work in some profiles. It'd be cool if I had a snippet for "Build Project" that had CMD syntax vs PowerShell syntax.

Idk exactly _how_ to solve that problem, but yeah, definitely something to consider

Comment (carlos-zamora) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:575]: This might actually be a really important thing too. I'm thinking about some nuances between CMD, PowerShell, and Linux. `git` commands are generally easy to copy/paste between the two, but the second you start talking about escaping characters and stuff like that, this can be annoying and difficult.

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 23]
```diff
@@ -470,16 +528,32 @@ their own workflows.
...
   [`clone_all_repos_in_org.yaml`](https://github.com/warpdotdev/workflows/blob/main/specs/git/clone_all_repos_in_org.yaml).
   We may want to straight up just seamlessly support that syntax as well.
-  Converting them to WT-compatible json is fairly trivial [[1](#footnote-1)].
-  Furthermore, the commands are all licensed under Apache 2.0, which means they
-  can be easily consumed by other OSS projects and shared with other developers.
-  This leads us to the next future consideration:
+  * Converting them to WT-compatible json is fairly trivial [[2](#footnote-2)].
+    * We may want to consider supporting YAML like this for `wt import`, ala [#10083]
+    * Similarly, we could import the YAML in the settings UI in a fashion similar
+      to how we import color schemes:
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:534]: We can/should probably mention the idea of importing .vscode/tasks.json here too. It'd be particularly cool if we could do that automatically since VS Code is also a first-party product.

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
Blocking over:
- [ ] Command Palette discussion (like, adding in the discussion points)
- [ ] discussion on .wt.json being next to the settings.json (see crazy idea below)

Crazy idea: what if we hard commit to snippets _only_ being in .wt.json files?

Here's what I mean:
- .wt.json files in the same directory as settings.json will be automatically loaded
- (Migration) `sendInput` actions in settings.json will be moved to the local .wt.json file
- Users can save .wt.json files from a separate repo to their local directory (besides the settings.json) so that they always have access to them
- `saveSnippet`/`x-save` saves the snippet to the .wt.json

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:283]: (Circling back to this) I'm thinking about the scenario where the .wt.json was improperly written. We'll try to load it then display some kind of warning to the user. Let's assume the user/author then goes into the .wt.json to fix it. We'll probably want to reload it. Otherwise, the dev loop for that would require the user to (1) know that they have to update their settings to re-load the .wt.json or (2) be a bit confused about why their changes to .wt.json aren't being propagated.

File: doc/specs/#1595 - Suggestions UI/Snippets.md (hunk 23 above)
Comment (zadjii-msft) [reply doc/specs/#1595 - Suggestions UI/Snippets.md:534]: I put that in the future considerations already 😉

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 24]
```diff
@@ -0,0 +1,622 @@
...
+
+We'll also want to engage our security partners to see if there's anything extra
+we'll need to do to ensure we're securely parsing random JSON files that we
+don't own.
+
+</td></tr>
+
+</table>
+
+### Other potential issues
```
Comment (carlos-zamora) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:454]: How does this work with `ssh`? Or is that something we don't have to worry about?

**Comment by zadjii-msft** [comment]
> Crazy idea: what if we hard commit to snippets only being in .wt.json files

Meh? I dunno, I'm not sure how valuable that is beyond just having a `.wt.json` that you stick in your dotfiles repo, then always stick in the root of the filesystem. Or just leaving then in the settings file. Just kinda feels like churn. I feel like all the user's settings they own should probably just stay in the one settings file we've already got. 

I'd be open to promoting `sendInput` from just an `action` to just a general `snippets` property in the `settings.json` itself. That's not too crazy.

**Review by carlos-zamora — APPROVED** [review]
Thanks for bearing with me. Excited for Snippets! 🎉

File: doc/specs/#1595 - Suggestions UI/Snippets.md [hunk 25]
```diff
@@ -455,7 +493,7 @@ and is going to try and redirect the _output of `wt`_ to `foo.txt`.
   `startingDirectory` (regardless of shell integration being enabled)
 * [ ] [#5790] - profile specific actions
 * [ ] [#12857] Ability to save selected text as a `sendInput` action
-* [ ] [#12861] Re-evaluate showing some sort of "ghost text or other preview for snippets
+* [x] [#12861] Re-evaluate showing some sort of "ghost text or other preview for snippets
```
Comment (DHowett) [inline doc/specs/#1595 - Suggestions UI/Snippets.md:496]:
```suggestion
* [x] [#12861] Re-evaluate showing some sort of "ghost text" or other preview for snippets
```

**Review by DHowett — APPROVED** [review]
R+ let's GOOOO

**Review by PankajBhojwani — APPROVED** [review]
_nervous excitement_

**Comment by zadjii-msft** [comment]
image
image

yOu CaNt UsE gItHuB aS aN iMaGe HoSt
